File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/columns/MulticolCloneGeometry.swift
//
//  MulticolCloneGeometry.swift
//  StyleEngine/columns — wave-46 lane Y3.
//
//  Pure fragment geometry for css-break-3 §5.2 `box-decoration-break:
//  clone` inside a multicol fragmentainer — the CLONE twin of
//  FragmentGeometry's slice S-table, and the iOS TWIN of the Android
//  runtime's MulticolCloneGeometry.kt. The shared K-table (full-fragment
//  rows K1/K2/K3/K7) is pinned in MulticolCloneGeometryTests.swift and
//  the Android MulticolCloneGeometryTest.kt with the SAME expected
//  fragments (native-pair parity gate).
//
//  The model: §5.2 — "each box fragment is independently wrapped with the
//  border, padding, and margin … the border-radius … applied to each
//  fragment independently. The background is drawn independently in each
//  fragment." So, unlike slice, every fragment re-spends the child's own
//  block-axis bands (border + padding) on its decoration and the CONTENT
//  advances by `H − bands` per fragment; and every fragment paints the
//  child as a COMPLETE box of the fragment's size, with NO block-axis
//  translate (the same box again, not a scrolled band of one tall box).
//
//  PLATFORM DIVERGENCE, BY DESIGN (documented in both pin suites): a
//  SHORT last fragment (leftover content + bands < H). iOS re-renders the
//  child per fragment (ComponentRenderer.multicolFragmentRow composes one
//  ComponentHost per fragment, and the plan hands it a child whose
//  declared block-size is the fragment's), so the short fragment is ONE
//  exact entry of its own height. Compose can measure its single child
//  only once (at H) and approximates the short fragment as two clipped
//  bands of that H-tall paint — see MulticolCloneGeometry.kt. Full
//  fragments are byte-identical across the pair.
//

// CoreGraphics for the CGRect/CGSize fragment shapes; Foundation for ceil.
import CoreGraphics
import Foundation

// Namespacing enum — all members static + pure (MulticolMath pattern).
enum MulticolCloneGeometry {

    /// The child's own block-axis decoration bands in px — what every
    /// clone fragment wraps its content slice with (css-break-3 §5.2).
    struct Bands: Equatable {
        /// border-top + padding-top (block-start band under horizontal-tb).
        let blockStartPx: Double
        /// border-bottom + padding-bottom (block-end band).
        let blockEndPx: Double
        /// Both bands — what the unfragmented layout spends exactly once.
        var totalPx: Double { blockStartPx + blockEndPx }
    }

    /// One clone fragment: the shared clip/translate shape the renderer's
    /// fragment row consumes, plus the fragment's own border-box block-size
    /// (the size the child is re-rendered at for this column).
    struct CloneFragment: Equatable {
        /// Column rect clip + inline-only translate (x_i, 0).
        let fragment: FragmentGeometry.Fragment
        /// The fragment's border-box block-size: H for every full fragment,
        /// `leftover content + bands` for a short last one.
        let boxBlockSizePx: Double
    }

    /// Clone fragment list for a child whose UNFRAGMENTED border-box
    /// block-size is `childBlockSize` (C = content + both bands) inside a
    /// fragmentainer of block-size `columnBlockSize` (H), used column width
    /// W, gap G and used count N.
    ///
    /// Nil when NO content fits a fragment (H ≤ bands) or there is no
    /// fragmentainer (H ≤ 0): the clone model has no answer there and the
    /// caller falls back to slice with a logOnce. Fragment count
    /// F = ceil(content / (H − bands)), at least 1, capped at N exactly
    /// like the slice table (the column-fill:auto overflow cap — the tail
    /// never gets a column; a capped remainder renders as a full H box).
    static func cloneFragments(childBlockSize: Double,
                               columnBlockSize: Double,
                               bands: Bands,
                               columnWidth: Double,
                               gapPx: Double,
                               columnCount: Int) -> [CloneFragment]? {
        // Defensive floors mirroring FragmentGeometry: count is a positive
        // <integer> (css-multicol §3.2), gaps non-negative (css-align §8),
        // a negative width clips to nothing, bands cannot be negative
        // (CSS 2.1 §8.4/§8.5 — padding and border widths are ≥ 0).
        let n = max(1, columnCount)
        let g = max(0, gapPx)
        let w = max(0, columnWidth)
        let startBand = max(0, bands.blockStartPx)
        let endBand = max(0, bands.blockEndPx)
        let h = columnBlockSize
        // Per-fragment CONTENT capacity: the fragmentainer minus the bands
        // every clone fragment re-spends on its own decoration (§5.2).
        let capacity = h - startBand - endBand
        // No fragmentainer, or bands alone fill it: clone is undefined
        // (every fragment would be pure decoration) — nil, the caller's
        // slice fallback + logOnce.
        guard h > 0, capacity > 0 else { return nil }
        // The content block-size: the unfragmented box minus ONE pair of
        // bands (the unfragmented layout wraps the content exactly once).
        let content = max(0, childBlockSize - startBand - endBand)
        // At least one fragment (an empty clone box still owns column 0),
        // at most N — the slice table's identical overflow cap.
        let count = min(n, max(1, Int(ceil(content / capacity))))
        return (0..<count).map { i in
            // Inline origin of column i — exactly i gaps precede column i.
            let x = Double(i) * (w + g)
            // Content left for THIS fragment: every fragment before the
            // last spends a full capacity, the last takes the remainder.
            // When the N cap truncated the list the remainder exceeds the
            // capacity and the fragment is a full H box (overflow clipped).
            let remaining = content - Double(i) * capacity
            // The fragment's border-box block-size: remainder + its own
            // bands, never more than the fragmentainer.
            let boxH = min(h, remaining + startBand + endBand)
            // Clip = the fragment's own rect in column i; translate = the
            // inline shift ONLY — the child is re-rendered at boxH, so no
            // block-axis scroll exists (unlike the slice table's −i·H).
            return CloneFragment(
                fragment: FragmentGeometry.Fragment(
                    columnIndex: i,
                    clipRect: CGRect(x: x, y: 0, width: w, height: boxH),
                    translate: CGSize(width: x, height: 0)),
                boxBlockSizePx: boxH)
        }
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/columns/MulticolCloneGeometry.kt
// Pure fragment geometry for css-break-3 §5.4 `box-decoration-break: clone`
// inside a multicol fragmentainer — the CLONE twin of FragmentGeometry's
// slice S-table. No Compose imports on purpose: the JUnit suite pins this
// math on the plain JVM, and the identical K-table is pinned on the iOS
// runtime (MulticolCloneGeometry.swift) so both natives clone byte-
// identically. Wave-46 lane Y3.
package com.styleconverter.runtime.columns

/**
 * Where each fragment of an over-tall `box-decoration-break: clone` child
 * lands, per css-break-3 §5.4: "each box fragment is independently wrapped
 * with the border, padding, and margin … the border-radius … applied to
 * each fragment independently. The background is drawn independently in
 * each fragment."
 *
 * The slice S-table lays the child out ONCE at its full block-size C and
 * scrolls band i into column i (translate −i·H). Clone is different in two
 * ways, and both fall out of ONE measuring trick the Compose and iOS
 * consumers share: the child is laid out at the FRAGMENT's block-size
 * (H for every full fragment) so its own decoration — borders, radius,
 * backgrounds, no-repeat images — paints as one complete box per fragment.
 *
 *  1. Capacity: every fragment spends `blockStartPx + blockEndPx` of its H
 *     on its OWN decoration bands (border + padding, §5.2 "independently
 *     wrapped"), so the content advances by `H − bands` per fragment, not
 *     by H. WPT css-break borders-008: 240px of content, H = 100, 10px
 *     borders → 80px per fragment → 3 fragments, each a full 100px circle.
 *  2. Replay: fragment i shows the H-tall clone paint with NO block
 *     translate (translateY = 0 — the same box, again). A SHORT last
 *     fragment (leftover content + bands < H) is composed from two bands
 *     of that same H-tall paint: its top rows verbatim, and its block-end
 *     decoration band taken from the BOTTOM of the paint (translate
 *     `boxH − H`), so the bottom border / bottom corners land at the
 *     fragment's true block-end edge. (iOS re-renders the child at the
 *     short size instead and gets an exact last fragment; the band split
 *     is the Compose single-measure approximation, documented in
 *     MulticolCloneMeasure.)
 *
 * Horizontal-tb only, like the slice table: vertical writing modes bail
 * (blocked-platform) before reaching this function.
 */
object MulticolCloneGeometry {

    /**
     * The child's own block-axis decoration bands, in px — what a clone
     * fragment wraps its content slice with (css-break-3 §5.4).
     */
    data class Bands(
        /** border-top + padding-top (block-start band under horizontal-tb). */
        val blockStartPx: Int,
        /** border-bottom + padding-bottom (block-end band). */
        val blockEndPx: Int,
        /**
         * The rows of the H-tall clone paint that must be copied to a
         * SHORT last fragment's block-end edge: `max(blockEndPx, the
         * largest bottom corner's vertical radius)` — a bottom corner arc
         * extends above the border band, so the band must be at least as
         * tall as the arc for the corners to rejoin (borders-008's circles).
         */
        val blockEndBandPx: Int
    )

    /**
     * Clone fragment list for a child whose UNFRAGMENTED border-box
     * block-size is [childBlockSizePx] (C, the intrinsic probe's answer —
     * content + both bands) inside a fragmentainer of block-size
     * [columnBlockSizePx] (H), used column width [columnWidthPx] (W), gap
     * [columnGapPx] (G) and used count [columnCount] (N).
     *
     * Returns null when NO content fits a fragment (H ≤ bands): the clone
     * model has no answer there — callers fall back to slice and log.
     * Fragment count F = ceil(content / (H − bands)), at least 1, capped at
     * N exactly like the slice table (the column-fill:auto overflow cap —
     * the tail simply never gets a column). Entries are in DRAW order:
     * full fragments contribute one entry; a short last fragment two (its
     * top rows, then its block-end band).
     */
    fun cloneFragments(
        childBlockSizePx: Int,
        columnBlockSizePx: Int,
        bands: Bands,
        columnWidthPx: Int,
        columnGapPx: Int,
        columnCount: Int
    ): List<FragmentGeometry.Fragment>? {
        // Defensive floors mirroring FragmentGeometry: count is a positive
        // <integer> (css-multicol §3.2), gaps non-negative (css-align §8),
        // a negative width clips to nothing, bands cannot be negative
        // (CSS 2.1 §8.4/§8.5 — padding and border widths are ≥ 0).
        val count = maxOf(1, columnCount)
        val gap = maxOf(0, columnGapPx)
        val width = maxOf(0, columnWidthPx)
        val startBand = maxOf(0, bands.blockStartPx)
        val endBand = maxOf(0, bands.blockEndPx)
        val h = columnBlockSizePx
        // Per-fragment CONTENT capacity: the fragmentainer minus the bands
        // every clone fragment re-spends on its own decoration (§5.2).
        val capacity = h - startBand - endBand
        // No fragmentainer, or bands alone fill it: clone is undefined
        // (every fragment would be pure decoration, content never
        // advances) — null, the caller's slice fallback + log.
        if (h <= 0 || capacity <= 0) return null
        // The content block-size: the unfragmented box minus ONE pair of
        // bands (the unfragmented layout wraps the content exactly once).
        val content = maxOf(0, childBlockSizePx - startBand - endBand)
        // ceil(content / capacity) in Long space so a pathological child
        // near Int.MAX_VALUE cannot wrap the ceiling addition negative.
        val rawCount = (content.toLong() + capacity - 1) / capacity
        // At least one fragment (an empty clone box still owns column 0),
        // at most N — the slice table's identical overflow cap.
        val fragmentCount = rawCount.coerceIn(1L, count.toLong()).toInt()
        // Build in draw order; a short last fragment appends two entries.
        val out = ArrayList<FragmentGeometry.Fragment>(fragmentCount + 1)
        for (i in 0 until fragmentCount) {
            // Inline origin of column i — exactly i gaps precede column i.
            val x = i * (width + gap)
            // Content left for THIS fragment; every fragment before the last
            // spends a full capacity, the last takes the remainder. When the
            // N cap truncated the list the remainder exceeds the capacity
            // and the fragment is a full H box (the overflow is clipped).
            val remaining = content - i.toLong() * capacity
            // The fragment's border-box block-size: remainder + its own
            // bands, never more than the fragmentainer.
            val boxH = minOf(h.toLong(), remaining + startBand + endBand).toInt()
            if (boxH >= h) {
                // A full fragment: the whole H-tall clone paint, untranslated
                // in the block axis (the same complete box, again).
                out.add(FragmentGeometry.Fragment(i, x, 0, width, h, x, 0))
            } else {
                // A SHORT last fragment. Its block-end band (border +
                // padding, widened to the bottom corner radius) comes from
                // the BOTTOM of the H-tall paint so the bottom border and
                // corner arcs sit at the fragment's real block-end edge;
                // the rows above it are the paint's own top rows.
                val band = bands.blockEndBandPx.coerceIn(0, boxH)
                val topRows = boxH - band
                // Top rows: clip [0, topRows), no translate.
                if (topRows > 0) {
                    out.add(FragmentGeometry.Fragment(i, x, 0, width, topRows, x, 0))
                }
                // Block-end band: clip [topRows, boxH), paint shifted UP by
                // (H − boxH) so the paint's bottom `band` rows land there.
                if (band > 0) {
                    out.add(FragmentGeometry.Fragment(i, x, topRows, width, band, x, boxH - h))
                }
            }
        }
        return out
    }
}
